nx/guides/advanced/complex_fft.livemd

```diff
@@ -0,0 +1,297 @@
+t = Nx.stack([t_one,Nx.complex(1,0)])
+
+Nx.abs(t)
+```
+
+We also have the imaginary number $i$ defined within `Nx.Constants`. Once imported, we can use it:
+
+```elixir
+i = Nx.Constants.i()
+
+Nx.add(1 , i)
+```
+
+It can be useful within a numerical function, defined using `defn`.
+
+For example, we can do:
+
+```elixir
+defmodule Example do
+  import Nx.Defn
+  import Nx.Constants, only: [i: 0]
+
+  defn rotate(z) do
+     i() * z
+  end
+end
+
+Nx.equal(Example.rotate(one), Nx.complex(-1,1))
+```
+
+## Discrete Fourrier Transform
+
+Suppose we have a periodic signal, and we want to sample it to understand the signal.
+
+We will run a Fast Fourrier Transform. It transforms a time-domain function into the frequency domain. Let's see how.
+
+<https://en.wikipedia.org/wiki/Discrete_Fourier_transform>
+
+<!-- livebook:{"break_markdown":true} -->
+
+### Build the signal
+
+Let's build a signal that we want to understand.
+
+It will be the sum of two sinusoidal signals, one at 5Hz, and one at 20Hz with the corresponding amplitudes (1, 0.5).
+
+$f(t) = \sin(2\pi*5*t) + \frac12 \sin(2\pi*20*t)$
+
+Suppose we can sample at `fs=50Hz` (meaning 50 samples per second) and our aquisition time is `duration = 1s`.
+
+We build a time series of `fs` points equally spaced with the given `duration` interval with `Nx.linspace`.
+
+For each value of this serie (the discrete time $t$), we will calculate the "hidden" signal $f(t)$ in the module below. This will give us the sampling.
+
+```elixir
+defmodule Signal do
+  import Nx.Defn
+  import Nx.Constants, only: [pi: 0]
+
+  defn source(t) do
+    f1 = 5; f2 = 20;
+    Nx.sin(2 * pi() * f1 * t ) + 1/2 * Nx.sin(2 * pi() * f2 * t)
+  end
+
+  defn sample(opts) do
+    fs = opts[:fs]
+    duration = opts[:duration]
+    sampling = Nx.linspace(0, duration, n: fs, endpoint: false, type: {:f, 32})
+    source(sampling)
+  end
+end
+```
+
+We sample our signal at fs=50Hz during 1s:
+
+```elixir
+fs = 50; duration= 1
+
+sample = Signal.sample(fs: fs, duration: 1)
+```
+
+### Analyse the signal with DFT
+
+The DFT algorithm will **approximates the original signal**. It gives us the following information:
+
+> The number at the index $i$ of the DFT results gives an approximation of the amplitude and phase of the sampled signal at the frequency $i$.
+
+In other words, doing `Nx.fft(sample)` returns a list of numbers indexed by the frequency.
+
+These are complex numbers as they carry the amplitude and phase information.
+
+```elixir
+dft = Nx.fft(sample)
+```
+
+Our frequency resolution - the accuracy of the DFT algorithm - is limited by the Nyquist rate, which is half the sampling rate, thus 25 Hz.
+
+Check [Nyquist](https://en.wikipedia.org/wiki/Nyquist_rate) to read more.
+
+We will therefor limit our study points to the indices below this threshold value by slicing the tensor, thus retain the 25 first ones.
+
+We are interested in the amplitudes only here. This is calculated by taking the norm of each number (`Nx.abs`).
```

Comment:
```suggestion
We will limit our study points to the first half of the `dft` because it is symmetrical on the upper half.
The phase doesn't really matter to us, so we'll use `Nx.abs` to obtain the absolute values at each point.
```

frequency resolution in general refers to the frequency delta between frequency bins, not the maximum representable unaliased bandwidth (which is related to the nyquist frequency/nyquist rate)

Reply:
ah, I see, yes of course, I used a wrong wording.  But yes, keep this short.